pages/ref/api-math.mdx

```diff
@@ -0,0 +1,245 @@
+        //                               ↑     ↑              ↑
+        //                               |     |              public_key, stored in our contract
+        //                               |     signature, obtained from the received message
+        //                               hash, calculated using the data from the received message
+        // ... follow-up logic ...
+    }
+}
+```
+
+## checkDataSignature
+
+```tact
+fun checkDataSignature(data: Slice, signature: Slice, public_key: Int): Bool;
+```
+
+Checks the [Ed25519][ed] `signature` of the `data` using a `public_key`, similar to [`checkSignature(){:tact}`](#checksignature). If the bit length of `data` is not divisible by $8$, this functions throws an error with [exit code 9](/book/exit-codes#9): `Cell underflow`. Verification itself is being done indirectly: on a [SHA-256][sha2] hash of the `data`.
+
+Returns `true{:tact}` if the signature is valid, `false{:tact}` otherwise.
+
+Usage example:
+
+```tact
+let data: Slice = ...;
+let signature: Slice = ...;
+let publicKey: Int = ...;
+
+let check: Bool = checkSignature(data, signature, publicKey);
+```
+
+## sha256
+
+```tact
+fun sha256(data: Slice): Int;
+fun sha256(data: String): Int;
+```
+
+Computes and returns the [SHA-256](https://en.wikipedia.org/wiki/SHA-2#Hash_standard) hash as an $256$-bit unsigned [`Int{:tact}`][int] from a passed [`Slice{:tact}`][p] or [`String{:tact}`][p] `data`.
+
+In case `data` is a [`String{:tact}`][p] it should have a number of bits divisible by $8$, and in case it's a [`Slice{:tact}`][p] it must **also** have no references (i.e. only up to 1023 bits of data in total).
+
+This function tries to resolve constant string values in [compile-time](/ref/api-comptime) whenever possible.
+
+Usage examples:
+
+```tact
+sha256(beginCell().asSlice());
+sha256("Hello, world!"); // will be resolved in compile-time
+sha256(someVariableElsewhere); // will try to resolve at compile-time,
+                               // and fallback to run-time evaluation
+```
+
+## min
+
+```tact
+fun min(x: Int, y: Int): Int;
+```
+
+Computes and returns the [minimum](https://en.wikipedia.org/wiki/Maximum_and_minimum) of two [`Int{:tact}`][int] values `x` and `y`.
+
+Usage examples:
+
+```tact
+min(1, 2);        // 1
+min(2, 2);        // 2
+min(007, 3);      // 3
+min(0x45, 3_0_0); // 69, nice
+//  ↑     ↑
+//  69    300
+```
+
+## max
+
+```tact
+fun max(x: Int, y: Int): Int;
+```
+
+Computes and returns the [maximum](https://en.wikipedia.org/wiki/Maximum_and_minimum) of two [`Int{:tact}`][int] values `x` and `y`.
+
+Usage examples:
+
+```tact
+max(1, 2);        // 2
+max(2, 2);        // 2
+max(007, 3);      // 7
+max(0x45, 3_0_0); // 300
+//  ↑     ↑
+//  69    300
+```
+
+## abs
+
+```tact
+fun abs(x: Int): Int
+```
+
+Computes and returns the [absolute value](https://en.wikipedia.org/wiki/Absolute_value) of the [`Int{:tact}`][int] value `x`.
```

Comment:
what happens if i run `abs` on the smallest `Int`?

Reply:
@anton-trunov Hmm, can't test it for some reason — the lowest number I'm able to get in Tact seems to be $-2^{256} + 1$, not $-2^{256}$. Creating the $-2^{256}$ throws exit code 4.

![image](https://github.com/tact-lang/tact-docs/assets/68142933/dd5066a8-6f85-4617-9657-2cb8d33b375f)

Using explicit literals doesn't work too, $-115792089237316195423570985008687907853269984665640564039457584007913129639936$ ($-2^{256}$) deemed invalid, while $-115792089237316195423570985008687907853269984665640564039457584007913129639935$ ($-2^{256} + 1$) works.

Seems like it's a bug of Tact, not TON.